### DISCUSSION TITLE
New-ish user wrote a clarification example for ORM 1-to-Many.  Is this useful to anyone?

### DISCUSSION THREAD

**Discussion by taigit** [sqlalchemy/sqlalchemy#7643] [Usage Questions]
I am beginning my ORM definition and looking at the ORM docs for 1-to-Many.
I wrote up this simple elaboration of 1-to-Many and thought it might be useful to others:

https://gist.github.com/taigit/21b2b1d04b3ac7d6208531ce7a5db95b

[sqla one to many.txt](https://github.com/sqlalchemy/sqlalchemy/files/7966389/sqla.one.to.many.txt)

**Comment by CaselIT** [comment]
Hi,

Thanks for posting. Feel free to open a PR if you find ways to improve the docs

**Comment by zzzeek** [comment]
well we can't just change one example, we'd need to change hundreds of examples everywhere, and I'm not sure I like the Hungarian notation as a new standard; its verbose and I find it to be distracting.  We'll be moving towards having real type annotations in all our examples so I think that will handle making it clear what datatype the various variables are storing.

**Reply by taigit** [reply]
Agreed.  This is not intended to be the "standard format" for examples.  

It's useful as:  let's look at one example in more detail and be explicit about it.   Then the other examples will be clearer.   Esp for beginners.

Maybe I'll just make it a stackoverflow question and answer and google can point people to it...

**Comment by taigit** [comment]
Now I have read and annotated the 1-to-1 relationship definition too.   It's actually different from the docs:

I added the  unique constraint that the "tip" in the docs refers to -- it seems like the unique constraint is part of the recommended mechanism for implementing 1-to-1.  Is that correct?

The gist link above has been updated to include 2 relationships.

**Reply by zzzeek** [reply]
it's likely that a unique constraint is a good idea but we dont want to in any way indicate that it's required.  Also, I dont want to encourage users to just stick "unique=True" on columns without understanding what that means, which is why the discussion of it points to a whole new section that I'd want the user to read.  adding unique=True has big implications for how database migrations are produced, how metadata should be configured, etc.    so no, I dont want users just flipping the flag on and forgetting about it because then they're going to come back with complaints about naming conventions, migrations, all of that.

**Reply by taigit** [reply]
What you say makes sense, but only after I spent 1+ hours understanding that example and tip...

And, actually, your explanation here about it is much more clear than the actual docs.  eg: I only understand your concerns after reading your comments.  

Yeah, I know, writing docs is hard and dull and painful...
For me, I would rather have your causal comments that you just typed be put in the docs, than have them missing from the docs.  Your casual comments are actually much easier to read than the docs (but maybe that's because I already studied the docs for 1+ hours).

**Reply by zzzeek** [reply]
i can attest that "spoken word" docs are unfortunately  ineffective as they lead to a lot of ambiguities and are also difficult for foreign language speakers.     in this case, the "tip" box links to the "unique constraint" concept, encouraging the user to please read that section.  I can add a phrase like "please read this link before using this flag" but again I dont know if that general way of documenting can scale.   it's very important that the style of language and the way things are explained be very consistent throughout all documentation.

**Reply by zzzeek** [reply]
here's an example of some docs where I was speaking off the cuff and a random word choice "discard" caused confusion: https://github.com/sqlalchemy/sqlalchemy/discussions/7687    it's such a big deal when you use verbs and nouns, that they all follow a consistent pattern

**Reply by taigit** [reply]
Wow, your SQLAlchemy world has so many [complex] concepts and they they interact  with the "real world"...  can be very confusing, yet it's important to describe the system.   It's a huge job!    It requires precision wording for the various concepts...

For me, I have just come across this distinction for "foreign keys":
`   Column(String, ForeignKey(  "" )  `
vs
`   relationship( "sqla_class", foreign_keys=""  )  `

I resolve it by putting in my comments: the second one can be thought of as:
`local_foreign_key_ref`
and then add more explanation...  so I can remember it later...

For me, your "spoken word" docs are excellent for understanding the overall concept and thought process -- they are very accessible and actually educational!  Then, when there is a precision coding question I will refer to the precise docs but it takes a lot more effort (for both the reader and writer).

**Comment by taigit** [comment]
OK, you have the best  view of the needs for your documentation.  I just want to add my vote saying your spoken word docs are really helpful.  The place for that is probably stackoverflow, I guess.
In fact, I'm sure you know but it's worth repeating: the official sqlalchemy docs are really a subset of the "real docs".  
the real docs are:    google search:  `sqlalchemy <topic>`

As such, I've posted my templates on stackoverflow.
I added your cautions about unique constraint to the comments.

I also added a 3rd category that I'm hitting.  I'm sure you know about it, but us newbies can't figure it out that easily.  I call them  1-way bookkeeping properties

https://stackoverflow.com/a/70969305/1023033

Your SQLAlchemy project Rocks!

**Reply by zzzeek** [reply]
see that's like some new terminology that seems to refer to something new.  why not call them what they are called, "many-to-one" ?   that's the canonical name for something like this.    "1-way bookkeeping" sounds like some way that you think of it when you are deciding what kinds of fields to have but that seems beyond the scope of what's needed to configure a "many-to-one" relationship.

**Reply by taigit** [reply]
oh, I did not realize that's what   "many-to-one"   meant.

I do a google search and figuring out the answer takes quite a bit of digging.   Not a good sign...

I actually like my term better so I keep it around in my stack overflow answer.    However, I did add a note saying this is called  "many-to-one" .   Thus, I link to the term and explain it a bit also.  (explanations in google search results are quite poor).

Thanks for the catch!
